rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,475 @@
+
+```py
+import tvm
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]):
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]):
+    A[0] = 0
+
+tvm.ir.assert_structural_equal(func_a, func_b)
+```
+
+The current error message indicates what the difference was, but not where it
+occurred.  This can sometimes be inferred from a stack trace, but becomes
+increasingly difficult with larger IR graphs.
+
+```
+ValueError: StructuralEqual check failed, caused by lhs:
+1
+and rhs:
+8
+```
+
+TUNIP should enable individual utilities and IR passes to have error messages
+directing the user to exact locations in the IR representation.
+
+```
+ValueError: StructuralEqual check failed, first delta highlighted below
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+```
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section introduces the design philosophy of the printer, and demonstrates
+the proposed user-facing APIs where users means IR developers.
+
+## Two-Stage Translation
+
+Traditionally in TVM stack, printing is a single-stage process. The printer
+assumes certain syntax of the target language, and therefore, so far there are
+3 different printers all for TIR: ReprPrinter, TIRTextPrinter,
+TVMScriptPrinter.
+
+We extend the idea of the existing Doc class at
+[src/printer/doc.h#L67](https://github.com/apache/tvm/blob/main/src/printer/doc.h#L67)
+to allow better consistency and scalability. An IR, which could
+be TIR, Relax or any other ones developed by third-party vendors, is first
+translated to an intermediate Doc node tree, and then the Doc tree is mapped to a target
+language, for example, Python, C++ IRBuilder API, or Rust.
+
+**Stage 1 [TVM IR => Doc]**. On the first stage, the printer needs to take care
+of translating a TVM IR to Doc tree. As an example, `tir.For` is translated to
+`ForDoc` without having to worry about the underlying language. Note that some
+complicated nodes in TVM IR, for example, `PrimFunc`, could be translated to
+multiple IR elements, including `FunctionDoc` and a few `StmtDoc`.
+
+During the translation from IR to Doc tree, it is possible that some statement
+influences the syntax of its children or vice verse, especially for syntactic
+sugars and declaring undefined variables in IR fragment printing. Therefore, a
+generic data structure `Frame` is introduced to allow retrieval and
+manipulation the relevant context information.
+
+**Stage 2. [Doc => target language]**. On the second stage, Doc tree is then
+honestly translated to the target language in text format. For example, when
+the target language is python, `ForDoc` is translated to python’s for loop
+syntax:
+
+```python
+for ... in ...:
+  ...
+```
+
+When the target language becomes python IRBuilder, `ForDoc` is translated to:
```

Comment:
just curious, do we have an example of using a different target language other than Python?

Reply:
Yes. Example target languages could be python/C++ IRBuilder we will develop later on for TVMScript metaprogramming (which is not part of the RFC though). Conceptually it's something like:

```python
with Builder():
  with T.prim_func(name="main"):
    A = T.arg(T.Buffer((128, 128), "float32"))
    B = T.arg(T.Buffer((128, 128), "float32"))
    with T.grid(128, 128, 128) as i, j, k:
      with T.block(name="block"):
        vi = T.axis.S(128, i)
        vj = T.axis.S(128, j)
        vk = T.axis.R(128, k)
```

The initial point of having two-stage translation though is to match the existing behavior, not anything new